### PR TITLE
Add unit tests for NWWaitingHandler, closes #589

### LINKED ISSUE #589: Add unit tests for `NWWaitingHandler`
[When we added `NWWaitingHandler`](https://github.com/swift-server/async-http-client/blob/main/Sources/AsyncHTTPClient/NIOTransportServices/NWWaitingHandler.swift) in https://github.com/swift-server/async-http-client/pull/588 we didn’t add a unit test for the `NWWaitingHandler`. This is because we need to make an enhancement to `NIOTransportServices` to make the `WaitingForConnectivity` event initialiser public. This is tracked as a separate issue in https://github.com/apple/swift-nio-transport-services/issues/147 and must be resolved first.

The unit test can be implemented by using an `EmbeddedChannel`  and sending an `NIOTSNetworkEvents.WaitingForConnectivity` through `fireUserInboundEventTriggered(_:)`. An example usage of `EmbeddedChannel` and `fireUserInboundEventTriggered(_:)`  can be found here:
https://github.com/swift-server/async-http-client/blob/3fcd67061f17d105282701c8ebd32467e458ceb6/Tests/AsyncHTTPClientTests/TLSEventsHandlerTests.swift#L24-L32

This should be straight forward and a good first starter issue. If you have more questions please don't hesitate to ask here.

### PR DESCRIPTION
**Pull request by natikgadzhi**
**Motivation:**

Closes #589. Since we already have a public initializer for `NIOTSNetworkEvents.WaitingForConnectivity`, we should add unit tests for the handler now that it's straightforward.

I'm here to learn — first patch in `async-http-client`.

**Modifications:**

The tests are in their own file `Tests/NWWaitingHandlerTests.swift`.

### CODE REVIEW

**Comment by natikgadzhi** [comment]
@dnadoba, hello there! I'm just learning, and that felt like a good issue to start with. How does that look?

**Comment by natikgadzhi** [comment]
@weissi, perhaps you would like to +1 a few unit tests? Am I doing this right?

**Review by Lukasa — CHANGES_REQUESTED** [review]
This requires lifting the minimum version of swift-nio-transport-services to 1.13.0, can you please add that change?

**Comment by natikgadzhi** [comment]
@Lukasa, thanks for reviewing this, and for calling out the dependency mismatch. 

`swift build && swift test` works locally now. `scripts/soundness.sh` has formatting mismatch in a bunch of files with the default `swiftformat` from brew, but that's unrelated to this PR.

File: Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        XCTAssertEqual(requester.transientError, NWError.dns(1))
+    }
+
+    func testWaitingHandlerDoesNotInvokeWaitingForConnectionOnUnrelatedErrors() {
+        let requester = MockRequester()
+        let waitingEventHandler = NWWaitingHandler(requester: requester, connectionID: 1)
+        let embedded = EmbeddedChannel(handlers: [waitingEventHandler])
+        embedded.pipeline.fireUserInboundEventTriggered(NIOTSNetworkEvents.BetterPathAvailable())
+
+        XCTAssertFalse(requester.waitingForConnectivityCalled, "Should not call .waitingForConnectitivy on unrelated events")
```
Comment (dnadoba) [inline Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift:67]:
```suggestion
        XCTAssertFalse(requester.waitingForConnectivityCalled, "Should not call .waitingForConnectivity on unrelated events")
```

File: Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    func testWaitingHandlerPassesTheEventDownTheContext() {
+        let requester = MockRequester()
+        let waitingEventHandler = NWWaitingHandler(requester: requester, connectionID: 1)
+        let tlsEventsHandler = TLSEventsHandler(deadline: nil)
+        let embedded = EmbeddedChannel(handlers: [waitingEventHandler, tlsEventsHandler])
+
+        embedded.pipeline.fireErrorCaught(NIOSSLError.handshakeFailed(BoringSSLError.wantConnect))
+        XCTAssertThrowsError(try XCTUnwrap(tlsEventsHandler.tlsEstablishedFuture).wait()) {
+            XCTAssertEqual($0 as? NIOSSLError, .handshakeFailed(BoringSSLError.wantConnect))
```
Comment (dnadoba) [inline Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift:78]: This prints the original error if it is not a `NIOSSLError`. Custom function in AHC so you couldn't possible know about that.
```suggestion
            XCTAssertEqualTypeAndValue($0, NIOSSLError.handshakeFailed(BoringSSLError.wantConnect))
```

**Review by dnadoba — CHANGES_REQUESTED** [review]
Looks really good! Two nits and this should be ready to be merged.
Sorry for the delayed review and thanks for tackling this!

**Review by Lukasa — APPROVED** [review]

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Comment by Lukasa** [comment]
A few formatting issues in the test:

```diff
diff --git a/Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift b/Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift
index ebbb851..9a1fa72 100644
--- a/Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift
+++ b/Tests/AsyncHTTPClientTests/NWWaitingHandlerTests.swift
@@ -16,28 +16,24 @@
 @testable import AsyncHTTPClient
 import Network
 import NIOCore
-import NIOTransportServices
 import NIOEmbedded
 import NIOSSL
+import NIOTransportServices
 import XCTest
 
 @available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
 class NWWaitingHandlerTests: XCTestCase {
-
     class MockRequester: HTTPConnectionRequester {
         var waitingForConnectivityCalled = false
-        var connectionID: AsyncHTTPClient.HTTPConnectionPool.Connection.ID? = nil
-        var transientError: NWError? = nil
+        var connectionID: AsyncHTTPClient.HTTPConnectionPool.Connection.ID?
+        var transientError: NWError?
+
+        func http1ConnectionCreated(_: AsyncHTTPClient.HTTP1Connection) {}
+
+        func http2ConnectionCreated(_: AsyncHTTPClient.HTTP2Connection, maximumStreams: Int) {}
+
+        func failedToCreateHTTPConnection(_: AsyncHTTPClient.HTTPConnectionPool.Connection.ID, error: Error) {}
 
-        func http1ConnectionCreated(_: AsyncHTTPClient.HTTP1Connection) {
-        }
-        
-        func http2ConnectionCreated(_: AsyncHTTPClient.HTTP2Connection, maximumStreams: Int) {
-        }
-        
-        func failedToCreateHTTPConnection(_: AsyncHTTPClient.HTTPConnectionPool.Connection.ID, error: Error) {
-        }
-        
         func waitingForConnectivity(_ connectionID: AsyncHTTPClient.HTTPConnectionPool.Connection.ID, error: Error) {
             self.waitingForConnectivityCalled = true
             self.connectionID = connectionID
@@ -50,9 +46,9 @@ class NWWaitingHandlerTests: XCTestCase {
         let connectionID: AsyncHTTPClient.HTTPConnectionPool.Connection.ID = 1
         let waitingEventHandler = NWWaitingHandler(requester: requester, connectionID: connectionID)
         let embedded = EmbeddedChannel(handlers: [waitingEventHandler])
-        
+
         embedded.pipeline.fireUserInboundEventTriggered(NIOTSNetworkEvents.WaitingForConnectivity(transientError: .dns(1)))
-        
+
         XCTAssertTrue(requester.waitingForConnectivityCalled, "Expected the handler to invoke .waitingForConnectivity on the requester")
         XCTAssertEqual(requester.connectionID, connectionID, "Expected the handler to pass connectionID to requester")
         XCTAssertEqual(requester.transientError, NWError.dns(1))
@@ -81,4 +77,3 @@ class NWWaitingHandlerTests: XCTestCase {
 }
 
 #endif
-
```

**Comment by dnadoba** [comment]
FYI: we use [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) in version `0.48.8` which is defined [here](https://github.com/swift-server/async-http-client/blob/62c06d47c8d21c91335e9f8998589e4ce31411e6/docker/Dockerfile#L31) if you want to run it locally yourself to auto format your code.

**Comment by natikgadzhi** [comment]
@dnadoba, thank you for the tip on where to lookup the specific swiftFormat version! Soundness checks don't verify that it's installed (and don't fail if it's not), and it looks like the swiftformat version might not be the same across all SSWG repositories. 

Is there a community best practice on how to manage multiple repos and versions of swiftformat? The brew formula does not work with versioned installs, but it's straightforward to checkout a git tag and rebuild — should I just do that for now?

**Comment by natikgadzhi** [comment]
- Merged `main` into this, so Transport version got booped to `1.19`.
- Cleaned up formatting
- Committed the suggestions from @dnadoba (thank you!)

Local tests are happy.

**Comment by dnadoba** [comment]
> Soundness checks don't verify that it's installed (and don't fail if it's not)

That is unfortunate. We have a set of [nicer scripts in OpenAPI](https://github.com/apple/swift-openapi-generator/tree/main/scripts) that catch this kind of issue though. We eventually want to have the same set of soundness checks across all of our repos.

> Is there a community best practice on how to manage multiple repos and versions of swiftformat?

Not to my knowledge. I build them from source and luckily rarely need to switch versions for the projects I work on.

**Comment by dnadoba** [comment]
@swift-server-bot test this please

**Comment by natikgadzhi** [comment]
When I pushed up my first PR to swift-metrics, I noticed that, and did this: https://github.com/apple/swift-metrics/pull/134 — but, it's not all that useful in a one-off pull request in one repository, like you mentioned, @dnadoba.

There's also this https://github.com/apple/swift-nio/pull/2242 attempt to implement soundness check as a SwiftPM plugin, and I think I saw another take on that as well — and a discussion that the internal team have not yet aligned on what's the best spot to put the universal soundness checks in. 

I'm pretty sure that once we align on how to do it, the rollout to all repositories will be easy, but if I can help out in any way — I'd be happy to. Seems like that might be tricky to do from outside of Apple, at least at the alignment stage. :eyes:

**Pull request merged**